[SYSTEMML-1532] Python launch script for spark-submit

krishnakalyan3 opened the pull request:
This script has been test on 
- OSX
- Ubuntu
- Windows 7

Reference : systemml-standalone.py and sparkDML.sh

akchinSTC commented:
Can an Admin verify this patch?

krishnakalyan3 commented:
@nakul02 Could you please review the work so far?. Its comparatively more robust. 

Some missing features below. (I could add them incase required)
```
   --stats                     Monitor and report caching/recompilation statistics
   --explain                   Explain plan (runtime)
   --explain2 <string>         Explain plan (hops, runtime, recompile_hops, recompile_runtime)
``` 

Only if every thing looks okay, I would prefer to work on adding support for the windows script.

nakul02 commented:
This is great work!
Complete all the options. Do not name the parameters `--explain` and `--explain 2`. Since you are using the argparse module, you should be able to do this in a better way. 
Make sure that all of systemml's arguments are covered through your options.

nakul02 commented on bin/systemml-spark-submit.py:
```diff
@@ -0,0 +1,131 @@
+parser.add_argument('--executor-memory', default='2G', help='Memory per executor', metavar='\b')
+parser.add_argument('--executor-cores', default='1', help='Number of cores', metavar='\b')
+parser.add_argument('--conf', default='', help='Configuration settings', metavar='\b')
+parser.add_argument('-f', required=True, help='DML script file name', metavar='\b')
```
There are 2 distinct set of arguments - one set to spark (master, executors...) and one set to systemml (nvargs, args, explain, stats...). In the python script, separate them out visually, add a newline with relevant comments.

krishnakalyan3 commented:
@nakul02 thank you for the review. I will incorporate these changes shortly.

krishnakalyan3 commented:
@nakul02 could you please let me know how I could obtain all input arguments for systemml jar file?. I am unable to locate the main java class that takes in arguments like (nvargs, args, explain, stats...).

nakul02 commented:
[Here](https://github.com/apache/incubator-systemml/blob/master/src/main/java/org/apache/sysml/api/DMLScript.java#L361)

nakul02 commented on bin/systemml-spark-submit.py:
```diff
@@ -18,40 +18,42 @@
+                                                   'heavy hitter <count> is 10 unless overridden')
+cparser.add_argument('-explain', default='runtime', help='explains plan levels can be hops, runtime, '
+                                                         'recompile_hops, recompile_runtime', metavar='')
+cparser.add_argument('-exe', default='hybrid_spark', help='System-ML backend (e.g spark, spark-hybrid)', metavar='')
```
Did you mean '-exec'

krishnakalyan3 replied:
In line number `126` I have to make a dictionary call `args.exe`. Unfortunately `args.exec` seems to give me a problem.

![image](https://cloud.githubusercontent.com/assets/4648995/26279861/9f5ed514-3d8d-11e7-8b8c-1511d9c76c5e.png)

krishnakalyan3 commented:
@nakul02, I would really appreciate another review. I have added more arguments except the ones stated below.  

```
Missing  `clean`
Missing `gpu`
Missing `debug`
Missing `python`
Missing `s` (specified script string to execute directly)
```

If we want to add these options, I would need to write around 5 `if else` conditions to check if these arguments are present. (Concurrently, I am looking for more elegant solution).

Changed `exec` to `exe` in the code as exec is a `python` reserved system keyword

deroneriksson commented:
cc @nakul02

nakul02 commented:
I am having trouble using this on Linux. What are the minimum arguments I'd need to give it?
This is what I tried:
```
bin/systemml-spark-submit.py --master local matmult.dml
```

Error I got:
```
usage: systemml-spark-submit.py [--help] [--master] [--driver-memory]
                                [--num-executors] [--executor-memory]
                                [--executor-cores] [--conf  [...]] -nvargs
                                NVARGS [NVARGS ...] [-args  [...]] [-config]
                                [-stats STATS] [-explain] [-exec] -f
systemml-spark-submit.py: error: argument -nvargs is required
```

My `$SPARK_HOME` is set correctly.
matmult.dml:
```R
A = rand(rows=10, cols=1000)
B = rand(rows=1000, cols=10)
C = A%*% B

print(toString(C))
```

krishnakalyan3 commented:
@nakul02 `'-nvargs', required=True` and `'-f', required=True` are compulsory arguments. If you want this behaviour changed, Please let me know.

nakul02 commented:
This is good. What platforms other than mac have you tested this on?

krishnakalyan3 commented:
@nakul02 I will be testing on ubuntu today. I have started working on configuring my windows instance to test this.

krishnakalyan3 commented:
I have trouble building system-ml from source on windows

- Installed java jdk8
- installed maven
- ran `mvn clean package`
(Error below)

![image](https://cloud.githubusercontent.com/assets/4648995/26370452/cfd41ba2-3fc5-11e7-8381-dacb729a471d.png)

ping @nakul02 @niketanpansare any thoughts?

niketanpansare commented:
Are you using maven2 ? If yes, can you please update to maven3: http://apache.github.io/incubator-systemml/engine-dev-guide

krishnakalyan3 commented:
@niketanpansare I am using maven3 :(

niketanpansare commented:
Interesting, the image in the above comment specifies https://repo.maven.org/maven2.

krishnakalyan3 commented:
Yes, I am not sure why.

![image](https://cloud.githubusercontent.com/assets/4648995/26377765/ade004ce-3fdf-11e7-911b-770b21818679.png)

niketanpansare commented:
Can you go to your `~/.m2/repository/org` directory and delete jcuda directory ? Then try `mvn package` again.

krishnakalyan3 commented:
@niketanpansare thank you for the suggestions. root cause 32bit windows. Currently I have other problems like the one below:

![image](https://cloud.githubusercontent.com/assets/4648995/26607435/7652e622-4564-11e7-953e-bcd750438884.png)
spark-submit / spark-shell does not work on windows.

- I have configured JAVA_HOME, SPARK_HOME, HADOOP_HOME, SYSTEMML_HOME
- HADOOP_HOME contain `winutils`

Any thoughts / suggestions?.

krishnakalyan3 commented:
ping @nakul02 @niketanpansare, Could you please review.
 
Tested 
- on Mac/Windows/Ubuntu 

`Program Files` in the environment variable causing the whitespace issue. 
Thanks for the debug help @nakul02.

nakul02 commented:
I get this warning when running on my machine -
```
Warning: Ignoring non-spark config property: /home/njindal/git/incubator-systemml/spark.driver.extraJavaOptions=-Dlog4j.configuration=file:/home/njindal/git/incubator-systemml/conf/log4j.properties.template
log4j:WARN No appenders could be found for logger (org.apache.hadoop.util.Shell).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
```

krishnakalyan3 commented:
@nakul02 fixed. 

There was a `windows backslash` problem.

nakul02 commented:
It shows me an explain even when I haven't specified it:

```
➜  incubator-systemml git:(master) ✗ bin/systemml-spark-submit.py -f matmult.dml               
================================================================================
Output dir: /home/njindal/git/incubator-systemml/temp
17/06/05 11:30:27 INFO api.DMLScript: BEGIN DML run 06/05/2017 11:30:27
17/06/05 11:30:27 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
17/06/05 11:30:27 INFO api.DMLScript: EXPLAIN (RUNTIME):
# Memory Budget local/remote = 3186MB/?MB/?MB/?MB
# Degree of Parallelism (vcores) local/remote = 8/?
PROGRAM ( size CP/SP = 10/0 )
--MAIN PROGRAM
----GENERIC (lines 1-8) [recompile=false]
------CP createvar _mVar0 scratch_space//_p6710_127.0.1.1//_t0/temp0 true MATRIX binaryblock 1000 30000 1000 1000 30000000 copy
------CP rand 1000 30000 1000 1000 0.0 1.0 1.0 -1 uniform 1.0 8 _mVar0.MATRIX.DOUBLE
------CP createvar _mVar1 scratch_space//_p6710_127.0.1.1//_t0/temp1 true MATRIX binaryblock 30000 1000 1000 1000 30000000 copy
------CP rand 30000 1000 1000 1000 0.0 1.0 1.0 -1 uniform 1.0 8 _mVar1.MATRIX.DOUBLE
------CP createvar _mVar2 scratch_space//_p6710_127.0.1.1//_t0/temp2 true MATRIX binaryblock 1000 1000 1000 1000 -1 copy
------CP ba+* _mVar0.MATRIX.DOUBLE _mVar1.MATRIX.DOUBLE _mVar2.MATRIX.DOUBLE 8
------CP rmvar _mVar0
------CP rmvar _mVar1
------CP write _mVar2.MATRIX.DOUBLE Vout.SCALAR.STRING.true textcell.SCALAR.STRING.true .SCALAR.STRING.true
------CP rmvar _mVar2

SystemML Statistics:
Total elapsed time:		4.649 sec.
Total compilation time:		0.584 sec.
Total execution time:		4.065 sec.
Number of compiled Spark inst:	0.
Number of executed Spark inst:	0.
Cache hits (Mem, WB, FS, HDFS):	2/0/0/0.
Cache writes (WB, FS, HDFS):	3/0/1.
Cache times (ACQr/m, RLS, EXP):	0.000/0.001/0.001/0.438 sec.
HOP DAGs recompiled (PRED, SB):	0/0.
HOP DAGs recompile time:	0.000 sec.
Spark ctx create time (lazy):	0.000 sec.
Spark trans counts (par,bc,col):0/0/0.
Spark trans times (par,bc,col):	0.000/0.000/0.000 secs.
Total JIT compile time:		1.944 sec.
Total JVM GC count:		6.
Total JVM GC time:		0.054 sec.
LibMatrixDNN dense count (conv/bwdF/bwdD/im2col/maxBwd):	0/0/0/0/0.
LibMatrixDNN sparse count (conv/bwdF/bwdD/im2col/maxBwd):	0/0/0/0/0.
LibMatrixDNN conv(im2col/matmult), bwdF (im2col/matmult), bwdD (col2im/matmult) time:	0.000/0.000/0.000/0.000/0.000/0.000 sec.
Heavy hitter instructions:
 #  Instruction  Time(s)  Count  GPU
 1  ba+*           3.152      1  
 2  rand           0.473      2  
 3  write          0.439      1  
 4  createvar      0.001      3  
 5  rmvar          0.000      3  

17/06/05 11:30:31 INFO api.DMLScript: END DML run 06/05/2017 11:30:31
```

krishnakalyan3 commented:
@nakul02 I have made explain as an conditional now. I had included explain as an default argument as per my observations [here](https://github.com/apache/incubator-systemml/blob/master/src/main/java/org/apache/sysml/api/DMLScript.java#L380)

krishnakalyan3 commented:
ping @nakul02, can I have another review?.
As per our earlier discussion, I have tested this script on the following platforms

- [x] OSX
- [x] Ubuntu
- [x] Windows

On windows, I tested this on `command` prompt and `Cygwin`. Known issue does not work in `Cygwin`.

nakul02 commented:
Still a bunch of problems. I would recommend testing against `systemml-standalone.py`. This script should behave like that.

```
➜  incubator-systemml git:(gpu_nntest_fix) ✗ bin/systemml-spark-submit.py -f matmult.dml -explain
usage: systemml-spark-submit.py [--help] [--master] [--driver-memory]
                                [--num-executors] [--executor-memory]
                                [--executor-cores] [--conf  [...]]
                                [-nvargs NVARGS [NVARGS ...]] [-args  [...]]
                                [-config] [-exec] [-explain] [-debug]
                                [-stats STATS] -f
systemml-spark-submit.py: error: argument -explain: expected one argument
➜  incubator-systemml git:(gpu_nntest_fix) ✗ bin/systemml-spark-submit.py -f matmult.dml -stats  
usage: systemml-spark-submit.py [--help] [--master] [--driver-memory]
                                [--num-executors] [--executor-memory]
                                [--executor-cores] [--conf  [...]]
                                [-nvargs NVARGS [NVARGS ...]] [-args  [...]]
                                [-config] [-exec] [-explain] [-debug]
                                [-stats STATS] -f
systemml-spark-submit.py: error: argument -stats: expected one argument
➜  incubator-systemml git:(gpu_nntest_fix) ✗ bin/systemml-spark-submit.py --master 'local[5]' --conf spark.driver.extraJavaOptions="-Xmx15g"
usage: systemml-spark-submit.py [--help] [--master] [--driver-memory]
                                [--num-executors] [--executor-memory]
                                [--executor-cores] [--conf  [...]]
                                [-nvargs NVARGS [NVARGS ...]] [-args  [...]]
                                [-config] [-exec] [-explain] [-debug]
                                [-stats STATS] -f
systemml-spark-submit.py: error: argument -f is required

➜  incubator-systemml git:(gpu_nntest_fix) ✗ bin/systemml-spark-submit.py --master 'local[5]' --conf spark.driver.extraJavaOptions="-Xmx15g"
usage: systemml-spark-submit.py [--help] [--master] [--driver-memory]
                                [--num-executors] [--executor-memory]
                                [--executor-cores] [--conf  [...]]
                                [-nvargs NVARGS [NVARGS ...]] [-args  [...]]
                                [-config] [-exec] [-explain] [-debug]
                                [-stats STATS] -f
systemml-spark-submit.py: error: argument -f is required
```

`explain`, `stats`, `gpu` accept an optional parameter. Your script requires a parameter, which is not the expected behavior. The `--conf` parameter doesn't work.

krishnakalyan3 commented:
@nakul02 Thanks for the review. I will use `systemml-standalone.py` to test against my script.

krishnakalyan3 commented:
Hello @nakul02, `--conf` parameter was indeed not working.

What is the expected behaviour of this command below?
```
bin/systemml-spark-submit.py --master 'local[5]' --conf spark.driver.extraJavaOptions="-Xmx15g"
```
I was under the impression that this would fail anyway since the corresponding dml script is not provided. 

Second question:
Is this error message useful or should I just remove it and keep it simple?. 

```
usage: systemml-spark-submit.py [--help] [--master] [--driver-memory]
                                [--num-executors] [--executor-memory]
                                [--executor-cores] [--conf  [...]]
                                [-nvargs NVARGS [NVARGS ...]] [-args  [...]]
                                [-config] [-exec] [-explain] [-debug]
                                [-stats STATS] -f
```

akchinSTC commented:
Can an Admin verify this patch?

krishnakalyan3 commented:
@nakul02 I have tested the following commands on my local system. Except the gpu flag everything works. I am not sure why I get `jcuda` error.

```
bin/systemml-spark-submit.py --master local[5] --conf spark.driver.maxResultSize="0" spark.akka.frameSize="128"  -f bin/matmul.dml
bin/systemml-spark-submit.py --master local[5]  -f bin/matmul.dml
bin/systemml-spark-submit.py --master local[5] -f bin/matmul.dml -explain
bin/systemml-spark-submit.py --master local[5] -f bin/matmul.dml -stats
bin/systemml-spark-submit.py --master local[5] -f bin/matmul.dml -gpu
```
Error trace when used with -gpu flag below
```
17/06/08 22:19:15 INFO api.DMLScript: END DML run 06/08/2017 22:19:15
Exception in thread "main" java.lang.NoClassDefFoundError: jcuda/runtime/JCuda
        at org.apache.sysml.runtime.instructions.gpu.context.GPUContextPool.initializeGPU(GPUContextPool.java:69)
        at org.apache.sysml.runtime.instructions.gpu.context.GPUContextPool.getFromPool(GPUContextPool.java:116)
        at org.apache.sysml.api.ScriptExecutorUtils.executeRuntimeProgram(ScriptExecutorUtils.java:86)
        at org.apache.sysml.api.DMLScript.execute(DMLScript.java:833)
        at org.apache.sysml.api.DMLScript.executeScript(DMLScript.java:506)
        at org.apache.sysml.api.DMLScript.main(DMLScript.java:233)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:738)
        at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
        at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.ClassNotFoundException: jcuda.runtime.JCuda
        at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
        at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 15 more
Failed to run SystemML. Exit code :256
```

nakul02 commented on bin/systemml-spark-submit.py:
```diff
@@ -0,0 +1,178 @@
+
+scripts_dir = join(project_root_dir, 'scripts')
+build_dir = join(project_root_dir, 'target')
+target_jars = join(build_dir, '*.jar')
```
You want to include 
${project_root_dir}/target/*.jar
and
${project_root_dir}/target/lib/*.jar

nakul02 replied:
Without this, `-gpu` fails.

akchinSTC commented:
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1584/

nakul02 commented:
Let me know when you've tested this on the 3 platforms and I shall merge.

krishnakalyan3 commented:
Hello @nakul02 
Tested on OSX, Ubuntu and Windows(cmd prompt) things work as expected.

akchinSTC commented:
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1610/

nakul02 commented:
Thanks @krishnakalyan3, I shall merge. 
We can do a future PR is we've missed out something.

krishnakalyan3 commented:
@nakul02 thank you for the reviews.
